app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,21 @@
+module Users
+  class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+    before_action :user_from_auth, only: :google_oauth2
+
+    def google_oauth2
+      if @user.persisted?
+        flash[:notice] = I18n.t 'devise.omniauth_callbacks.success', kind: 'Google'
+        sign_in_and_redirect @user, event: :authentication
+      else
+        session['devise.google_data'] = request.env['omniauth.auth'].except(:extra)
+        redirect_to new_user_registration_url, alert: @user.errors.full_messages.join("\n")
+      end
+    end
+
+    private
+
+    def user_from_auth
+      @user = User.from_omniauth(request.env['omniauth.auth'])
+    end
+  end
+end
```

Comment:
@marderer let's use one-line class naming:
```ruby
class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
...
end
```

it keeps more available space.
What do you think about adding rubocop rule for that?

Reply:
@marderer see how it can be refactored:
```ruby
class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
  def google_oauth2
    if user_from_auth.persisted?
      flash[:notice] = I18n.t('devise.omniauth_callbacks.success'), kind: 'Google'
      sign_in_and_redirect user_from_auth, event: :authentication
    else
      set_devise_google_data
      redirect_to new_user_registration_url, alert: user_from_auth_error_messages
    end
  end

  private

  def set_devise_google_data
    session['devise.google_data'] = request.env['omniauth.auth'].except(:extra)
  end

  def user_from_auth_error_messages
    user_from_auth.errors.full_messages.join("\n")
  end

  def user_from_auth
    @user_from_auth ||= User.from_omniauth(request.env['omniauth.auth'])
  end
end
```

Reply:
@marderer lets move user creation logic to a server like
```ruby
Users::Omniauth.new(request.env["omniauth.auth"]).get
```
and the service:
```ruby
app/services/users/omniauth.rb

class Users::Omniauth
  attr_reader :auth

  def initialize(auth)
    @auth = auth
  end

  def get
    create_user unless user.persisted?
    
    user
  end

  private

  delegate :provider, :uid, :info, to: :auth
  delegate :email, to: :info

  def create_user
    user.skip_confirmation!
    user.save
  end

  def user
    @_user ||= User.find_or_initialize_by(
      provider: provider, 
      uid: uid, 
      email: email, 
      password: Devise.friendly_token[0, 20]
    )
  end
end
```

So we can move this logic out of `User` model.

Reply:
@AleksSenkou i agree with you, but rubocop requires write 
module A
  class B
   ...
  end
end

Better disable this rule in rubocop?

Reply:
@AleksSenkou use service for auth is good idea, thanks
But I made some changes, because this don't work how i wanted

Reply:
@marderer there is cop in Rubocop for that
```ruby
Style/ClassAndModuleChildren:
  EnforcedStyle: compact
  Exclude:
    - 'config/**/*'
```